File: sdk-source/csharp/Aegis.Sdk/AegisException.cs
using System;

namespace Aegis.Sdk
{
    /// <summary>Every Aegis API failure surfaces as this exception.</summary>
    public class AegisException : Exception
    {
        /// <summary>Documented API error code, e.g. "hwid_mismatch".</summary>
        public string Code { get; }

        /// <summary>HTTP status code (0 for transport failures).</summary>
        public int StatusCode { get; }

        public AegisException(string code, string message, int statusCode = 0, Exception? inner = null)
            : base(message, inner)
        {
            Code = code;
            StatusCode = statusCode;
        }

        public bool IsAuthError => Code == "unauthorized" || Code == "invalid_credentials";
        public bool IsLicenseError => Code.StartsWith("license", StringComparison.Ordinal) || Code == "hwid_mismatch";
        public bool IsNetworkError => StatusCode == 0;

        public override string ToString() => $"AegisException({Code}, http {StatusCode}): {Message}";
    }
}

File: sdk-source/php/src/AegisException.php
<?php

declare(strict_types=1);

namespace Aegis;

/** Every Aegis API failure surfaces as this exception. */
class AegisException extends \RuntimeException
{
    public function __construct(
        public readonly string $errorCode,
        string $message,
        public readonly int $status = 0,
        ?\Throwable $previous = null,
    ) {
        parent::__construct($message, $status, $previous);
    }

    public function isNetworkError(): bool
    {
        return $this->status === 0;
    }

    public function isAuthError(): bool
    {
        return in_array($this->errorCode, ['unauthorized', 'invalid_credentials'], true);
    }

    public function isLicenseError(): bool
    {
        return str_starts_with($this->errorCode, 'license') || $this->errorCode === 'hwid_mismatch';
    }
}